File: 2016/Day14/main.py
import hashlib
import re
from functools import lru_cache

def calculate_md5(input_string):
  return hashlib.md5(input_string.encode('utf-8')).hexdigest()

def calculate_stretched_md5(input_string):
  hash_val = calculate_md5(input_string)
  for _ in range(2016):
    hash_val = calculate_md5(hash_val)
  return hash_val

TRIPLET_REGEX = re.compile(r'(.)\1\1')

def find_first_triplet_char(hash_str):
  match = TRIPLET_REGEX.search(hash_str)
  return match.group(1) if match else None

def has_quintuplet(hash_str, char):
  return (char * 5) in hash_str

def find_keys(salt, num_keys, hash_function):
  index = 0
  keys_found = 0
  @lru_cache(maxsize=2000)
  def get_hash(i):
    return hash_function(salt + str(i))

  while keys_found < num_keys:
    current_hash = get_hash(index)
    triplet_char = find_first_triplet_char(current_hash)

    if triplet_char:
      for i in range(index + 1, index + 1001):
        next_hash = get_hash(i)
        if has_quintuplet(next_hash, triplet_char):
          keys_found += 1
          if keys_found == num_keys:
            return index
          break

    index += 1

  return -1

if __name__ == "__main__":
  SALT = "zpqevtbw"
  NUM_KEYS_TO_FIND = 64

  print(f"Index of the 64th key: {find_keys(SALT, NUM_KEYS_TO_FIND, calculate_md5)}")
  print(f"Index of the 64th key with stretched hash: {find_keys(SALT, NUM_KEYS_TO_FIND, calculate_stretched_md5)}")

File: 2016/Day14/go/main.go
package main

import (
	"crypto/md5"
	"encoding/hex"
	"fmt"
	"strconv"
	"strings"
)

func md5Hash(input string) string {
	hasher := md5.New()
	hasher.Write([]byte(input))
	return hex.EncodeToString(hasher.Sum(nil))
}

func stretchedMd5Hash(input string) string {
	hash := md5Hash(input)
	for range 2016 {
		hash = md5Hash(hash)
	}
	return hash
}

func hasTriplet(hash string) (bool, byte) {
	for i := range len(hash) - 2 {
		if hash[i] == hash[i+1] && hash[i] == hash[i+2] {
			return true, hash[i]
		}
	}
	return false, 0
}

func hasQuintuplet(hash string, char byte) bool {
	quintuplet := strings.Repeat(string(char), 5)
	return strings.Contains(hash, quintuplet)
}

type hashFunction func(string) string

func findKeyIndex(salt string, hashFunc hashFunction) int {
	index := 0
	keysFound := 0
	hashes := make(map[int]string)

	getHash := func(idx int) string {
		if hash, ok := hashes[idx]; ok {
			return hash
		}
		computedHash := hashFunc(salt + strconv.Itoa(idx))
		hashes[idx] = computedHash
		return computedHash
	}

	for keysFound < 64 {
		currentHash := getHash(index)
		foundTriplet, tripletChar := hasTriplet(currentHash)

		if foundTriplet {
			for i := index + 1; i <= index+1000; i++ {
				nextHash := getHash(i)
				if hasQuintuplet(nextHash, tripletChar) {
					keysFound++
					break
				}
			}
		}

		index++
	}

	return index - 1
}

func main() {
	salt := "zpqevtbw"

	fmt.Printf("Index of the 64th key: %d\n", findKeyIndex(salt, md5Hash))
	fmt.Printf("Index of the 64th key with stretched hash: %d\n", findKeyIndex(salt, stretchedMd5Hash))
}
